src/runtime/xrt.cpp

```diff
@@ -0,0 +1,628 @@
+}  // namespace Runtime
+}  // namespace Halide
+
+using namespace Halide::Runtime::Internal;
+using namespace Halide::Runtime::Internal::XRT;
+
+extern "C" {
+
+WEAK int halide_xrt_acquire_context(void *user_context,
+                                    void **instance_ret,
+                                    void **adapter_ret,
+                                    void **device_ret,
+                                    bool create = true) {
+    halide_debug_assert(user_context, &thread_lock != nullptr);
+
+    halide_mutex_lock(&thread_lock);
+
+    if (create && (global_device == nullptr)) {
+        int status = create_xrt_context(user_context);
+        if (status != halide_error_code_success) {
+            halide_mutex_unlock(&thread_lock);
+            return status;
+        }
+    }
+
+    *instance_ret = global_instance;
+    *adapter_ret = global_adapter;
+    *device_ret = global_device;
+
+    return halide_error_code_success;
+}
+
+WEAK int halide_xrt_release_context(void *user_context) {
+    halide_mutex_unlock(&thread_lock);
+    return halide_error_code_success;
+}
+
+}  // extern "C" linkage
+
+namespace Halide {
+namespace Runtime {
+namespace Internal {
+namespace XRT {
+
+// Helper object to acquire and release the XRT context.
+class XRTContext {
+    void *user_context;
+
+public:
+    void *instance = nullptr;
+    void *adapter = nullptr;
+    xrtDeviceHandle device = nullptr;
+    void *queue = nullptr;
+
+    int error_code = 0;
+
+    ALWAYS_INLINE XRTContext(void *user_context)
+        : user_context(user_context) {
+        error_code = halide_xrt_acquire_context(
+            user_context, &instance, &adapter, &device);
+        if (error_code == halide_error_code_success) {
+            halide_start_clock(user_context);
+        }
+    }
+
+    ALWAYS_INLINE ~XRTContext() {
+        (void)halide_xrt_release_context(user_context);  // ignore errors
+    }
+};
+
+// XrtBufferHandle represents a device buffer
+struct XrtBufferHandle {
+    // If nullptr, it means not allocated yet
+    xrtBufferHandle handle;
+    size_t size;
+    bool copy_to_device_pending;
+};
+
+// XrtKernelState represents a loaded kernel on the device
+struct XrtKernelState {
+    xrtKernelHandle handle;
+};
+
+WEAK int create_xrt_context(void *user_context) {
+    unsigned int count = xclProbe();
+    debug(user_context) << "XRT: create_xrt_context: found: " << count << " devices\n";
+
+    if (count == 0) {
+        error(user_context) << "XRT: create_xrt_context: error: no devices were found\n";
+        return halide_error_code_gpu_device_error;
+    }
+
+    for (unsigned i = 0; i < count; i++) {
+        xrtDeviceHandle device = xrtDeviceOpen(i);
+        debug(user_context) << "XRT: create_xrt_context: xrtDeviceOpen: " << device << "\n";
+        if (device != nullptr) {
+            global_device = device;
+            return halide_error_code_success;
+        }
+    }
+
+    return halide_error_code_gpu_device_error;
```

Comment:
Error handling in the Halide runtime is (alas) hard to get right and easy to get wrong (it's needed an overhaul for a while but so far has not happened).

But, generally speaking, the Golden Rule is:
- If you call `error()` or `halide_error()` in the runtime code, you *must* return a nonzero error code from the Runtime API boundary.
- Similarly, if you return a nonzero error code, you must call `error()` or `halide_error()`. 

This is to allow for the fact that a given runtime environment might define  a `halide_error()` implementation that simply calls `abort()` or `exit()`... or it might define one that does nothing (except perhaps log a message), meaning that the result code is the only way for the caller to check for failure.